base/compiler/ssair/show.jl

```diff
@@ -37,69 +39,266 @@ function print_node(io::IO, idx, stmt, used, maxsize; color = true, print_typ=tr
             e = stmt.edges[i]
             v = !isassigned(stmt.values, i) ? "#undef" :
                 sprint() do io′
-                    print_ssa(io′, stmt.values[i])
+                    print_ssa(io′, stmt.values[i], argnames)
                 end
             "$e => $v"
         end
         Base.print(io, "φ ", '(', join(args, ", "), ')')
     elseif isa(stmt, PhiCNode)
-        Base.print(io, "φᶜ ", '(', join(map(x->sprint(print_ssa, x), stmt.values), ", "), ')')
+        Base.print(io, "φᶜ ", '(', join(map(x->sprint(print_ssa, x, argnames), stmt.values), ", "), ')')
     elseif isa(stmt, PiNode)
         Base.print(io, "π (")
-        print_ssa(io, stmt.val)
+        print_ssa(io, stmt.val, argnames)
         Base.print(io, ", ")
         if color
-            Base.printstyled(io, stmt.typ, color=:red)
+            Base.printstyled(io, stmt.typ, color=:cyan)
         else
             Base.print(io, stmt.typ)
         end
         Base.print(io, ")")
     elseif isa(stmt, UpsilonNode)
         Base.print(io, "ϒ (")
         isdefined(stmt, :val) ?
-            print_ssa(io, stmt.val) :
+            print_ssa(io, stmt.val, argnames) :
             Base.print(io, "#undef")
         Base.print(io, ")")
     elseif isa(stmt, ReturnNode)
         if !isdefined(stmt, :val)
             Base.print(io, "unreachable")
         else
             Base.print(io, "return ")
-            print_ssa(io, stmt.val)
+            print_ssa(io, stmt.val, argnames)
         end
     elseif isa(stmt, GotoIfNot)
         Base.print(io, "goto ", stmt.dest, " if not ")
-        print_ssa(io, stmt.cond)
+        print_ssa(io, stmt.cond, argnames)
     elseif isexpr(stmt, :call)
-        print_ssa(io, stmt.args[1])
+        print_ssa(io, stmt.args[1], argnames)
         Base.print(io, "(")
-        Base.print(io, join(map(arg->sprint(io->print_ssa(io, arg)), stmt.args[2:end]), ", "))
+        Base.print(io, join(map(arg->sprint(io->print_ssa(io, arg, argnames)), stmt.args[2:end]), ", "))
+        Base.print(io, ")")
+        if print_typ && stmt.typ !== Any
+            Base.print(io, "::$(stmt.typ)")
+        end
+    elseif isexpr(stmt, :invoke)
+        print(io, "invoke ")
+        linfo = stmt.args[1]
+        print_ssa(io, stmt.args[2], argnames)
+        Base.print(io, "(")
+        print_arg((typ, arg)) = sprint(io->(print_ssa(io, arg, argnames); print(io, "::$(typ)")))
+        Base.print(io, join(map(print_arg, zip(linfo.specTypes.parameters[2:end], stmt.args[3:end])), ", "))
```

Comment:
This code is very hard to parse. Why are you stringifying everything multiple times? At least skip the obvious copy:
```julia
Base.join(io, map(print_arg, zip(linfo.specTypes.parameters[2:end], stmt.args[3:end])), ", ")
```

`linfo.specTypes.parameters[:]` - this quantity doesn't always exist. use `fieldtype(linfo.specTypes, i)`.

As I've said before, using `zip` on `args` (in Inference) is usually a very bad idea (very, very slow, and poisons the dispatch, Tuple, and Type caches for later with too many entries). There's some helper functions in Core.Compiler (such as `_any`, `contains_is`, and `anymap`) which help ensure Inference compiles fast and small which should be used instead (and add more if you need them).

Reply:
(actually, don't use `fieldtype`, since doing this approximate transform at all will convey incorrect information about the invoke site)